BLUESPAWN-win-client/src/util/filesystem/FileSystem.cpp

```diff
@@ -0,0 +1,991 @@
+
+        bool success{ false };
+        HCRYPTPROV hash{};
+        switch(hashType) {
+        case HashType::SHA1_HASH: success = CryptCreateHash(hProv, CALG_SHA1, 0, 0, &hash); break;
+        case HashType::MD5_HASH: success = CryptCreateHash(hProv, CALG_MD5, 0, 0, &hash); break;
+        case HashType::SHA256_HASH: success = CryptCreateHash(hProv, CALG_SHA_256, 0, 0, &hash); break;
+        }
+
+        GenericWrapper<HCRYPTPROV> hHash{ hash, CryptDestroyHash };
+
+        if(!success) {
+            LOG_ERROR("Unable to create hash providers");
+            return std::nullopt;
+        }
+
+        if(!CryptHashData(hHash, data.GetAsPointer<const BYTE>(), data.GetSize(), 0)) {
+            LOG_ERROR("Failed to hash file " << FilePath << " (" << SYSTEM_ERROR << ")");
+            return std::nullopt;
+        }
+
+        DWORD cbHashSize = 0, dwCount = sizeof(DWORD);
+        if(!CryptGetHashParam(hHash, HP_HASHSIZE, (BYTE*) &cbHashSize, &dwCount, 0)) {
+            LOG_ERROR("Failed to read hash size for " << FilePath << " (" << SYSTEM_ERROR << ")");
+            return std::nullopt;
+        }
+
+        std::vector<BYTE> buffer(cbHashSize);
+        if(!CryptGetHashParam(hHash, HP_HASHVAL, reinterpret_cast<BYTE*>(buffer.data()), &cbHashSize, 0)) {
+            LOG_ERROR("Failed to read hash for " << FilePath << " (" << SYSTEM_ERROR << ")");
+            return std::nullopt;
+        }
+
+        std::wstringstream oss{};
+        for(auto& byte : buffer) {
+            oss.fill('0');
+            oss.width(2);
+            oss << std::hex << static_cast<const int>(byte);
+        }
+
+        return oss.str();
+    }
+
+    File::File(IN const std::wstring& path) : hFile{ nullptr } {
+        if(!path.length()) {
+            bFileExists = false;
+            bWriteAccess = false;
+            bReadAccess = false;
+            return;
+        }
+        FilePath = ExpandEnvStringsW(path);
+        LOG_VERBOSE(2, "Attempting to open file: " << FilePath << ".");
+        if(FilePath.at(0) == L'\\') {
+            UNICODE_STRING UnicodeName{ static_cast<USHORT>(FilePath.length() * 2),
+                                        static_cast<USHORT>(FilePath.length() * 2),
+                                        const_cast<PWCHAR>(FilePath.c_str()) };
+            OBJECT_ATTRIBUTES attributes{};
+            IO_STATUS_BLOCK IoStatus{};
+            InitializeObjectAttributes(&attributes, &UnicodeName, OBJ_KERNEL_HANDLE | OBJ_CASE_INSENSITIVE, nullptr,
+                                       nullptr);
+            NTSTATUS Status{ Linker::NtCreateFile(&hFile, GENERIC_READ | GENERIC_WRITE, &attributes, &IoStatus, nullptr,
+                                                  FILE_ATTRIBUTE_NORMAL, FILE_SHARE_READ, FILE_OPEN,
+                                                  FILE_SEQUENTIAL_ONLY, nullptr, 0) };
+            if(NT_SUCCESS(Status)) {
+                this->hFile = hFile;

```

Comment:
Codacy found an issue: [Redundant assignment of 'this->hFile' to itself.](https://app.codacy.com/manual/ION28/BLUESPAWN/pullRequest?prid=5982821)